cohortextractor/query_engines/mssql_lib.py

```diff
@@ -0,0 +1,339 @@
+            for n, query in enumerate(preparatory_queries):
+                log.info("Running query {n}/{len(queries)}")
+                connection.execute(query)
+            # Run the write to temporary table within an explicit transaction
+            # so we can't end up in a state where the table exists but is
+            # half-populated. We have to commit the existing implicit
+            # transaction before we can start the new one.
+            connection.commit()
+            with connection.begin():
+                log.info(f"Running final query and writing results to '{table_name}'")
+                connection.execute(write_query_to_table(table, select_query))
+                log.info(f"Creating '{key_column}' index on '{table_name}'")
+                connection.execute(create_index_for_table(table))
+                connection.commit()
+
+        # Here we pass back an iterator over the temporary table as the value
+        # of the context manager
+        yield fetch_table_in_batches(
+            connection, table, key_column, **batch_fetch_config
+        )
+
+        # If all goes well we clean up the temporary table; if there was an
+        # error then we'll leave it in place for next time
+        connection.execute(sqlalchemy.schema.DropTable(table, if_exists=True))
+        connection.commit()
+
+
+def fetch_table_in_batches(
+    connection,
+    table,
+    key_column,
+    batch_size=32000,
+    max_retries=2,
+    sleep=0.5,
+    reconnect_on_error=False,
+):
+    """
+    Returns an iterator over all the rows in a table by querying it in batches,
+    retrying automatically on errors.
+
+    Args:
+        connection: SQLAlchemy Connection object, or ReconnectableConnection
+            (see below)
+
+        table: SQLAlchemy Table object
+
+        key_column: name of a unique integer column in the results, used for paging
+
+        batch_size: how many results to fetch in each batch
+
+        max_retries: how many *sequential* failures to retry after
+
+        sleep: time in seconds to sleep between retries
+
+        reconnect_on_error: whether to close and reopen database connection
+            after each error; this is robust against more forms of failure (e.g
+            the underlying TCP connection being closed), but can't be used with
+            session-scoped temporary tables for obvious reasons. If this option
+            is supplied then `connection` must be a ReconnectableConnection
+            instance
+    """
+    if reconnect_on_error and not hasattr(connection, "reconnect"):
+        raise ValueError(
+            "Connection must be a ReconnectableConnection if "
+            "`reconnect_on_error` is used"
+        )
+
+    key = sqlalchemy.Column(key_column)
+    retries = 0
+    batch_count = 0
+    total_rows = 0
+    min_key = None
+
+    while True:
+        query = (
+            sqlalchemy.select("*").select_from(table).order_by(key).limit(batch_size)
+        )
```

Comment:
Unless there's an index on `key`, I think this might be accidentally quadratic.  Not sure how much of a problem this is.

Reply:
You're quite right. We do create the appropriate index on the temporary table before passing it through here. But I'll add a note to the docstring in case we end up using this function in other contexts.